### PR TITLE
Fix #1148: Show clearer error message for mismatched types in QueryBuilder

### LINKED ISSUE #1148: Clearer error messages for mismatched types
**Is your feature request related to a problem? Please describe.**
- Create a dataframe with a string column representing a numeric identifier
- Query the column using QueryBuilder, but mistakenly using an int in the filter

The user gets an error like,

```
File "/users/isys/glai/pyenvs/res-new/lib/python3.6/site-packages/arcticdb/version_store/_store.py", line 1686, in _read_dataframe
    return ReadResult(*self.version_store.read_dataframe_version(symbol, version_query, read_query, read_options))
arcticdb_ext.exceptions.InternalException: E_ASSERTION_FAILURE Cannot compare TD<type=UTF_DYNAMIC64, dim=0> to TD<type=UINT32, dim=0> (possible categorical?)
```

**Describe the solution you'd like**
- We should translate the message in to "user-space" a bit more. We can't expect users to translate UTF_DYNAMIC64 in to "Python string"
- We don't tell the user the affected column (slightly relates to #944 )

### PR DESCRIPTION
**Pull request by muhammadhamzasajjad**
#### Reference Issues/PRs
Fixes #1148.
previously if a type was mismatched in the QueryBuilder as follows
```python
df1 = pd.DataFrame({"col1": [1, 2, 3], "col2": [2, 3, 4], "col3": [4, 5, 6], "col_str": ["1", "2", "3"]})
sym = "symbol"
lib.write(sym, df1)
q = QueryBuilder()
q = q[q["col1" + 1]  == "3"]
lib.read(sym, query_builder=q)
```
it would show an unclear message with `UTF_DYNAMIC64` type shown for strings 
```
arcticdb_ext.exceptions.InternalException: E_ASSERTION_FAILURE Cannot compare TD<type=UTF_DYNAMIC64, dim=0> to TD<type=UINT32, dim=0> (possible categorical?)
```
Now the error is much clearer where column names are generated according to the query and STRING type is shown for strings
```
arcticdb_ext.exceptions.UserInputException: E_INVALID_USER_ARGUMENT Invalid comparison (col1 + 1) (TD<type=INT64, dim=0>) == "3" (TD<type=STRING, dim=0>)
```
For a more complex query like `q = q[1 + q["col1"] * q["col2"] - q["col3"] == q["col_str"]]` it will show column name `((1 + (col1 * col2)) - col3)` in the error message which allows the user to better understand the error.

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [x] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [x] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp [hunk 1]
```diff
@@ -273,20 +282,25 @@ VariantData visit_binary_comparator(const VariantData& left, const VariantData&
     }, left, right);
 }
 
+template <typename Func>
+inline std::string binary_operation_column_name(std::string_view left_column, Func& func, std::string_view right_column) {
```
Comment (vasil-pashov) [inline cpp/arcticdb/processing/operation_dispatch_binary.hpp:286]: The proper thing is to use universal reference
```suggestion
inline std::string binary_operation_column_name(std::string_view left_column, Func&& func, std::string_view right_column) {
```

File: cpp/arcticdb/processing/operation_types.hpp [hunk 2]
```diff
@@ -520,3 +520,182 @@ bool operator()(T t, const ankerl::unordered_dense::set<U>& u) const {
...
+};
+
+template<>
+struct formatter<arcticdb::EqualsOperator> {
+    template<typename ParseContext>
+    constexpr auto parse(ParseContext &ctx) { return ctx.begin(); }
+
+    template<typename FormatContext>
+    constexpr auto format(const arcticdb::EqualsOperator&, FormatContext &ctx) const {
+        return fmt::format_to(ctx.out(), "=");
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_types.hpp:620]: ==

File: cpp/arcticdb/processing/operation_types.hpp [hunk 3]
```diff
@@ -520,3 +520,182 @@ bool operator()(T t, const ankerl::unordered_dense::set<U>& u) const {
...
+};
+
+template<>
+struct formatter<arcticdb::IsInOperator> {
+    template<typename ParseContext>
+    constexpr auto parse(ParseContext &ctx) { return ctx.begin(); }
+
+    template<typename FormatContext>
+    constexpr auto format(const arcticdb::IsInOperator&, FormatContext &ctx) const {
+        return fmt::format_to(ctx.out(), "IN");
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_types.hpp:686]: `IS IN`

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp (hunk 1 above)
Comment (alexowens90) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp:286]: William had to make some changes in these files to get everything working with C++17 again, if you haven't already done so a rebase is a good idea

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp [hunk 4]
```diff
@@ -105,8 +105,10 @@ VariantData binary_membership(const ColumnWithStrings& column_with_strings, Valu
                     }
                 });
             } else {
-                util::raise_rte("Cannot check membership of {} in set of {} (possible categorical?)",
-                                column_with_strings.column_->type(), value_set.base_type());
+                util::raise_rte("Cannot check membership of {} {} in set of {} (possible categorical?)",
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_binary.hpp:108]: Please replace all of these `util::raise_rte` with `E_INVALID_USER_ARGUMENT`

File: cpp/arcticdb/processing/expression_node.hpp [hunk 5]
```diff
@@ -42,19 +43,23 @@ class StringPool;
 struct ColumnWithStrings {
     std::shared_ptr<Column> column_;
     const std::shared_ptr<StringPool> string_pool_;
+    ColumnName column_name_;
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.hpp:46]: We may as well store this as a `std::string`, we only need `ColumnName` when it is ambiguous whether the string is a column name, value, expression, etc

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp [hunk 6]
```diff
@@ -105,8 +105,10 @@ VariantData binary_membership(const ColumnWithStrings& column_with_strings, Valu
...
+                                column_with_strings.column_name_.value,
+                                get_user_friendly_type_string(column_with_strings.column_->type()),
+                                get_user_friendly_type_string(value_set.base_type()));
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_binary.hpp:111]: Here and in the `binary_comparator` equivalents, the operator does not appear?

File: cpp/arcticdb/processing/operation_dispatch_unary.hpp [hunk 7]
```diff
@@ -44,7 +44,7 @@ VariantData unary_operator(const Value& val, Func&& func) {
     details::visit_type(val.type().data_type(), [&](auto val_tag) {
         using type_info = ScalarTypeInfo<decltype(val_tag)>;
         if constexpr (!is_numeric_type(type_info::data_type)) {
-            util::raise_rte("Cannot perform arithmetic on {}", val.type());
+            util::raise_rte("Cannot perform arithmetic on {}", get_user_friendly_type_string(val.type()));
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_unary.hpp:47]: No operator in exception message

File: cpp/arcticdb/processing/operation_dispatch_unary.hpp [hunk 8]
```diff
@@ -56,36 +56,43 @@ VariantData unary_operator(const Value& val, Func&& func) {
...
                                                                                                      *output_column,
                                                                                                      [&func](auto input_value) -> TargetType {
                                                                                                          return func.apply(
                                                                                                                  input_value);
                                                                                                      });
         } else {
-            util::raise_rte("Cannot perform arithmetic on {}", col.type());
+            util::raise_rte("Cannot perform arithmetic on {} {}",
+                            col.column_name_.value,
+                            get_user_friendly_type_string(col.column_->type()));
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_unary.hpp:85]: No operator

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp [hunk 9]
```diff
@@ -353,24 +372,26 @@ VariantData binary_operator(const Column& col, const Value& val, Func&& func) {
                 constexpr auto output_data_type = data_type_from_raw_type<ReversedTargetType>();
                 output_column = std::make_unique<Column>(make_scalar_type(output_data_type), true);
                 Column::transform<typename col_type_info::TDT, ScalarTagType<DataTypeTag<output_data_type>>>(
-                        col,
+                        *(col.column_),
                         *output_column,
-                        [&func, raw_value](auto input_value) -> ReversedTargetType {
-                    return func.apply(raw_value, input_value);
+                        [&func, &col, &column_name, raw_value](auto input_value) -> ReversedTargetType {
+                            column_name = binary_operation_column_name(fmt::format("{}", raw_value), func, col.column_name_.value);
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_binary.hpp:378]: This cannot go here, this lambda is called for every value in the column

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp [hunk 10]
```diff
@@ -353,24 +372,26 @@ VariantData binary_operator(const Column& col, const Value& val, Func&& func) {
...
                 constexpr auto output_data_type = data_type_from_raw_type<TargetType>();
                 output_column = std::make_unique<Column>(make_scalar_type(output_data_type), true);
                 Column::transform<typename col_type_info::TDT, ScalarTagType<DataTypeTag<output_data_type>>>(
-                        col,
+                        *(col.column_),
                         *output_column,
-                        [&func, raw_value](auto input_value) -> TargetType {
-                    return func.apply(input_value, raw_value);
+                        [&func, &col, &column_name, raw_value](auto input_value) -> TargetType {
+                            column_name = binary_operation_column_name(col.column_name_.value, func, fmt::format("{}", raw_value));
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_binary.hpp:388]: Same as above

**Review by alexowens90 — CHANGES_REQUESTED** [review]

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp (hunk 1 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp:286]: > William had to make some changes in these files to get everything working with C++17 again, if you haven't already done so a rebase is a good idea

@alexowens90, I already did that. You can see [here](https://github.com/man-group/ArcticDB/commits/bug/query_type_mismatch_message/) that William's C++17 revert is included in this branch.

Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: cpp/arcticdb/processing/operation_types.hpp (hunk 2 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_types.hpp]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: cpp/arcticdb/processing/operation_types.hpp (hunk 3 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_types.hpp]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp (hunk 4 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: cpp/arcticdb/processing/expression_node.hpp (hunk 5 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/expression_node.hpp]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp (hunk 6 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp:112]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: cpp/arcticdb/processing/operation_dispatch_unary.hpp (hunk 7 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_dispatch_unary.hpp]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp (hunk 9 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp (hunk 10 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp]: Fixed in [a907db9](https://github.com/man-group/ArcticDB/pull/1557/commits/a907db94c4366371434d048c630f796fcbc7a60e)

File: python/tests/unit/arcticdb/version_store/test_query_builder.py [hunk 11]
```diff
@@ -268,6 +268,40 @@ def test_querybuilder_filter_then_filter(lmdb_version_store_tiny_segment):
     assert_frame_equal(expected, received)
 
 
+def test_df_query_wrong_type(lmdb_version_store_small_segment):
```
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_query_builder.py:271]: This should move to `test_filtering.py`

File: cpp/arcticdb/processing/operation_dispatch_unary.hpp (hunk 8 above)
Comment (alexowens90) [reply cpp/arcticdb/processing/operation_dispatch_unary.hpp:86]: Can reuse `unary_operation_column_name` here for more consistency:
```
"Cannot perform unary operation '{}' on column of type {}",
                            unary_operation_column_name(func, col.column_name_),
                            get_user_friendly_type_string(col.column_->type())
```

**Review by alexowens90 — CHANGES_REQUESTED** [review]

File: cpp/arcticdb/processing/operation_dispatch_unary.hpp [hunk 12]
```diff
@@ -94,36 +105,39 @@ VariantData visit_unary_operator(const VariantData& left, Func&& func) {
...
+    if (is_empty_type(col.column_->type().data_type()) || is_integer_type(col.column_->type().data_type())) {
         if constexpr (std::is_same_v<std::remove_reference_t<Func>, IsNullOperator>) {
-            return is_empty_type(col.type().data_type()) ? VariantData(FullResult{}) : VariantData(EmptyResult{});
+            return is_empty_type(col.column_->type().data_type()) ? VariantData(FullResult{}) : VariantData(EmptyResult{});
         } else if constexpr (std::is_same_v<std::remove_reference_t<Func>, NotNullOperator>) {
-            return is_empty_type(col.type().data_type()) ? VariantData(EmptyResult{}) : VariantData(FullResult{});
+            return is_empty_type(col.column_->type().data_type()) ? VariantData(EmptyResult{}) : VariantData(FullResult{});
         } else {
-            internal::raise<ErrorCode::E_ASSERTION_FAILURE>("Unexpected operator passed to unary_comparator");
+            user_input::raise<ErrorCode::E_INVALID_USER_ARGUMENT>("Unexpected operator passed to unary_comparator");
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_unary.hpp:122]: This one isn't a user input error, it should still be `E_ASSERTION_FAILURE`

File: cpp/arcticdb/processing/operation_dispatch_unary.hpp [hunk 13]
```diff
@@ -134,19 +148,20 @@ template<typename Func>
...
+                return transform_to_placeholder(unary_comparator(l, std::forward<decltype(func)>(func)));
             },
             [] (EmptyResult) -> VariantData {
                 if constexpr (std::is_same_v<std::remove_reference_t<Func>, IsNullOperator>) {
                     return FullResult{};
                 } else if constexpr (std::is_same_v<std::remove_reference_t<Func>, NotNullOperator>) {
                     return EmptyResult{};
                 } else {
-                    internal::raise<ErrorCode::E_ASSERTION_FAILURE>("Unexpected operator passed to visit unary_comparator");
+                    user_input::raise<ErrorCode::E_INVALID_USER_ARGUMENT>("Unexpected operator passed to visit unary_comparator");
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_unary.hpp:159]: This one isn't a user input error, it should still be `E_ASSERTION_FAILURE`

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp [hunk 14]
```diff
@@ -265,30 +282,50 @@ VariantData visit_binary_comparator(const VariantData& left, const VariantData&
...
+
+template <typename Func>
+inline std::string binary_operation_with_types_to_string(std::string_view left, const TypeDescriptor& type_left, Func&& func,
+                                                         std::string_view right, const TypeDescriptor& type_right,
+                                                         bool arguments_reversed = false) {
+    if (arguments_reversed) {
+        return fmt::format("{} ({}) {} {} ({})", right, get_user_friendly_type_string(type_right), func, left, get_user_friendly_type_string(type_left));
+    }
+    return fmt::format("{} ({}) {} {} ({})", left, get_user_friendly_type_string(type_left), func, right, get_user_friendly_type_string(type_right));
+}
```
Comment (alexowens90) [inline cpp/arcticdb/processing/operation_dispatch_binary.hpp:308]: Can you move these two functions to the top of the file? It is odd that they appear in amongst the main methods

File: cpp/arcticdb/processing/operation_dispatch_unary.hpp [hunk 15]
```diff
@@ -56,36 +58,45 @@ VariantData unary_operator(const Value& val, Func&& func) {
 }
 
 template <typename Func>
-VariantData unary_operator(const Column& col, Func&& func) {
+inline std::string unary_operation_column_name(Func& func, std::string_view column_name) {
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/processing/operation_dispatch_unary.hpp:61]: Also, put at the top of the file for the same reason as in `operation_dispatch_binary.hpp`

File: cpp/arcticdb/pipeline/value.hpp [hunk 16]
```diff
@@ -208,4 +208,27 @@ inline std::optional<std::string> ascii_to_padded_utf32(std::string_view str, si
...
+    template<typename ParseContext>
+    constexpr auto parse(ParseContext &ctx) { return ctx.begin(); }
+
+    template<typename FormatContext>
+    auto format(const arcticdb::Value& value, FormatContext& ctx) const {
+        if (is_sequence_type(value.type().data_type())) {
+            fmt::format_to(ctx.out(), "\"{}\"", std::string(*value.str_data(), value.len()));
+        }
+        else {
+            arcticdb::entity::details::visit_type(value.data_type_, [&](auto val_tag) {
```
Comment (alexowens90) [inline cpp/arcticdb/pipeline/value.hpp:225]: This is going to introduce an extra visitation over all of the types, but everywhere this is called is already inside a visitation over the value's type

**Review by alexowens90 — CHANGES_REQUESTED** [review]

**Review by alexowens90 — APPROVED** [review]

**Pull request merged**